### PR TITLE
feat(loadbalancer): implement failover policy #488

### PR DESCRIPTION
**Pull request by acoliver**
## Summary

Implements failover policy for load balancer profiles (#488), part of the Load Balancer Master Issue (#485).

This PR adds a failover strategy to the LoadBalancingProvider that tries backend profiles sequentially until one succeeds. When the primary backend fails, requests automatically failover to the next configured backend, providing high availability for multi-provider setups.

### Core Load Balancer Implementation
- **New failover strategy** in `LoadBalancingProvider.ts` with `executeWithFailover()` method
- **Two-level retry system**: per-backend retries before failing over to the next backend
- **Streaming-aware error handling**: Awaits first chunk before yielding to detect errors early in streaming responses
- **New error type**: `LoadBalancerFailoverError` thrown when all backends are exhausted

### CLI Changes
- **Policy parameter is now required**: `/profile save loadbalancer <name> <roundrobin|failover> <profile1> <profile2> [...]`
- **Schema updated** with `roundrobin` and `failover` literals for autocomplete support
- **Parser validates** policy parameter before saving profile

### New Ephemeral Settings for Failover Configuration
| Setting | Description | Default |
|---------|-------------|---------|
| `failover_retry_count` | Number of retries per backend before failover | `1` |
| `failover_retry_delay_ms` | Delay between retry attempts in milliseconds | `1000` |
| `failover_on_network_errors` | Whether to failover on network errors | `true` |
| `failover_status_codes` | HTTP status codes that trigger failover | `[429, 500, 502, 503, 504]` |

### Profile Application
- Maps `policy: 'failover'` in profile JSON to `strategy: 'failover'` in LoadBalancingProvider
- Extracts failover settings from ephemeral settings when creating provider

## Implementation Approach

This was implemented using **strict TDD methodology** with a 3-phase approach:

1. **RED Phase**: Wrote 48 tests first (all failing initially)
2. **GREEN Phase**: Implemented minimal code to pass all tests
3. **REFACTOR Phase**: Cleaned up code, fixed lint/type issues, ran full verification

The implementation followed a detailed plan documented in `project-plans/20251212issue488/PLAN.md`.

## Files Changed

| File | Changes |
|------|---------|
| `packages/core/src/providers/LoadBalancingProvider.ts` | Core failover implementation |
| `packages/core/src/providers/errors.ts` | New `LoadBalancerFailoverError` class |
| `packages/core/src/types/modelParams.ts` | Failover setting type definitions |
| `packages/cli/src/ui/commands/profileCommand.ts` | CLI schema and parsing for required policy |
| `packages/cli/src/runtime/profileApplication.ts` | Profile-to-provider mapping |
| `packages/core/src/providers/__tests__/LoadBalancingProvider.failover.test.ts` | 22 new tests |
| `packages/cli/src/ui/commands/__tests__/profileCommand.failover.test.ts` | 17 new tests |
| `packages/cli/src/runtime/__tests__/profileApplication.failover.test.ts` | 9 new tests |

## Test Plan

- [x] All 48 new failover tests pass
- [x] Full test suite passes (6588 tests)
- [x] `npm run lint:ci` passes (zero warnings)
- [x] `npm run typecheck` passes
- [x] `npm run build` and `npm run bundle` pass
- [x] Happy path validation with `syntheticfailover` profile

## Example Usage

```bash
# Create a failover load balancer profile
/profile save loadbalancer my-failover failover primary-profile backup-profile

# Configure failover behavior
/set failover_retry_count 2
/set failover_retry_delay_ms 500
/set failover_status_codes [429, 500, 502, 503]

# Load and use
/profile load my-failover
```

## Related Issues

- Implements #488
- Part of #485 (Load Balancer Master Issue)
- Documentation will be added in #778 (Phase 6)

## Summary by CodeRabbit

* **New Features**
  * Added a failover load‑balancing policy alongside round‑robin, selectable when saving load‑balancer profiles.
  * Failover includes per‑backend retries, configurable delays, and propagation of ephemeral settings to providers.
  * Enhanced error reporting aggregates failures and lists attempted backends on total failure.

* **Tests**
  * Added and updated comprehensive tests covering policy parsing, provider creation, failover execution, streaming behavior, and edge cases.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: packages/cli/src/ui/commands/__tests__/profileCommand.failover.test.ts [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+        'loadbalancer lb-name RoundRobin profile1 profile2',
+      );
+
+      expect(runtimeMocks.saveLoadBalancerProfile).toHaveBeenCalledTimes(1);
+      const savedProfile = runtimeMocks.saveLoadBalancerProfile.mock
+        .calls[0][1] as LoadBalancerProfile;
+
+      expect(savedProfile.policy).toBe('roundrobin');
+    });
+  });
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/cli/src/ui/commands/__tests__/profileCommand.failover.test.ts:136]: this was not desired an llm hallucination. we'd rather be explicit

File: project-plans/20251212issue488/PLAN.md [hunk 2]
```diff
@@ -0,0 +1,800 @@
+# Failover Policy Implementation Plan - Issue #488 (REVISED)
+
+**Issue**: #488
+**Related**: #485 (design), #486 (round-robin implementation - completed)
+**Date**: 2025-12-12
+**Plan**: PLAN-20251212issue488
+**Revision**: 2 - Fixed streaming, error handling, CLI parsing, and edge cases
```
[the remark this answers is not in the document]
Comment (acoliver) [reply project-plans/20251212issue488/PLAN.md:7]: I don't really want you to validate plan documents after implementation.

**Comment by acoliver** [comment]
## CodeRabbit Review Response

Reviewed all CodeRabbit suggestions. Here's what was done:

### Fixed ✅
- **Inconsistent error message** (line 341): Updated fallback usage message to include `<roundrobin|failover>` for consistency with other usage messages in the file. Commit: 269082805

### Skipped (User Decision) 🚫
- **Policy parameter should be optional**: Intentionally rejected - we want explicit policy selection, not implicit defaults
- **Plan document markdown linting**: Plan documents are implementation artifacts, not production code

### Skipped (Codebase Conventions) 📋
- **Defensive copy for ephemeralSettings**: Type system guarantees `ephemeralSettings` is always present (`Record<string, unknown>`, not optional). No other code in this codebase does defensive copies for this pattern.
- **Object.setPrototypeOf + immutable failures array**: None of the other 4 error classes in `errors.ts` use `Object.setPrototypeOf` or create defensive copies of arrays. Following existing conventions. The empty failures case is impossible since `executeWithFailover` only throws after attempting all backends.
- **Iterator simplification**: The current `for await (const chunk of { [Symbol.asyncIterator]: () => iterator })` pattern is idiomatic JavaScript for continuing to yield from a partially-consumed async iterator. The suggested alternative is more verbose without benefit.

### Already Correct ✓
- Token slicing (slice(4)) for policy parameter
- Schema structure for policy selection
- Case-insensitive policy validation
- Profile object construction

**Comment by acoliver** [comment]
## CodeRabbit Review Response (Round 2)

### Fixed ✅
- **Path separator validation for lb-name**: Added validation to reject load balancer profile names containing `/` or `\`, matching the existing validation for model profiles. Commit: 31a05e0fb

### Already Fixed ✓
- **Inconsistent error message**: Was fixed in previous commit 269082805

### Not Fixed 🚫

**Profile names with spaces (#3)**
- Would require implementing a quote-aware tokenizer across ALL parameters (lb-name, policy, profile1, profile2, etc.)
- Current implementation uses simple `split(/\s+/)` which is consistent
- This is a significant refactor beyond the failover feature scope
- Should be tracked as a separate enhancement issue if desired

**Defensive copy for ephemeralSettings (#4)**
- LoadBalancingProvider already handles undefined defensively: `this.config.lbProfileEphemeralSettings ?? {}`
- Type system provides guarantees (`Record<string, unknown>`)
- No practical benefit to additional copying

**Deep copy for LoadBalancerFailoverError.failures (#5)**
- Already marked `readonly` with `ReadonlyArray<T>` - TypeScript best practice
- Empty array case is impossible (validation requires ≥2 profiles)
- Deep copying adds runtime overhead without practical benefit

**Simplify remaining chunks iteration (#6)**
- Current pattern `for await (const chunk of { [Symbol.asyncIterator]: () => iterator })` is idiomatic JavaScript
- Standard approach for continuing iteration from partially-consumed async iterator
- Suggested manual loop is more verbose without benefit
